src/main/java/org/elasticsearch/index/translog/fs/FsTranslog.java

```diff
@@ -190,135 +238,162 @@ public long ramBytesUsed() {
+                try {
+                    IOUtils.close(old);
+                } catch (IOException e) {
+                    logger.error("failed to closed old translog [{}] (committed id [{}])", e, old, translogId);
                 }
             }
-            this.trans = type.create(shardId, id, new InternalChannelReference(location.resolve(getPath(id)), StandardOpenOption.READ, StandardOpenOption.WRITE, StandardOpenOption.CREATE_NEW), transientBufferSize);
-        } catch (IOException e) {
-            throw new TranslogException(shardId, "failed to create new translog file", e);
-        } finally {
-            rwl.writeLock().unlock();
         }
     }
 
     @Override
-    public void makeTransientCurrent() throws IOException {
-        FsTranslogFile old;
-        rwl.writeLock().lock();
-        try {
-            assert this.trans != null;
-            old = current;
-            this.current = this.trans;
-            this.trans = null;
-        } finally {
-            rwl.writeLock().unlock();
+    public long newTranslog() throws TranslogException, IOException {
+        try (ReleasableLock _ = writeLock.acquire()) {
+            final FsTranslogFile newFile = createTranslogFile();
+            final FsTranslogFile old = current;
+            current = newFile;
+            newFile.reuse(old);
+            FsChannelImmutableReader reader = old.immutableReader();
+            uncomittedTranslogs.add(reader);
+            for (FsView view : outstandingViews) {
+                view.onNewTranslog(old.immutableReader(), current.reader());
+            }
+            IOUtils.close(old);
+            logger.trace("current translog set to [{}]", current.translogId());
+            return current.translogId();
         }
-        old.close();
-        current.reuse(old);
     }
 
-    @Override
-    public void revertTransient() throws IOException {
-        rwl.writeLock().lock();
+    protected FsTranslogFile createTranslogFile() throws IOException {
+        FsTranslogFile newFile;
+        long size = Long.MAX_VALUE;
+        Path location = null;
+        for (Path file : locations) {
+            long currentFree = Files.getFileStore(file).getUsableSpace();
+            if (currentFree < size) {
+                size = currentFree;
+                location = file;
+            } else if (currentFree == size && ThreadLocalRandom.current().nextBoolean()) {
+                location = file;
+            }
+        }
         try {
-            final FsTranslogFile toClose = this.trans;
-            this.trans = null;
-            IOUtils.close(toClose);
-        } finally {
-            rwl.writeLock().unlock();
+            long id = idGenerator++;
+            newFile = type.create(shardId, id, new InternalChannelReference(id, location.resolve(getPath(id)), StandardOpenOption.READ, StandardOpenOption.WRITE, StandardOpenOption.CREATE_NEW), bufferSize);
```

Comment:
`getPath` should return a String filename, see https://github.com/elastic/elasticsearch/pull/10621